x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/FileOperatorUsersStore.java

```diff
@@ -0,0 +1,284 @@
+        return operatorUsersDescriptor;
+    }
+
+    static final class OperatorUsersDescriptor {
+        private final List<Group> groups;
+
+        private OperatorUsersDescriptor(List<Group> groups) {
+            this.groups = groups;
+        }
+
+        // Package private for tests
+        List<Group> getGroups() {
+            return groups;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+            OperatorUsersDescriptor that = (OperatorUsersDescriptor) o;
+            return groups.equals(that.groups);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(groups);
+        }
+    }
+
+    private static final OperatorUsersDescriptor EMPTY_OPERATOR_USERS_DESCRIPTOR = new OperatorUsersDescriptor(List.of());
+
+    static final class Group {
+        private static final Set<String> SINGLETON_REALM_TYPES = Set.of(
+            FileRealmSettings.TYPE, NativeRealmSettings.TYPE, ReservedRealm.TYPE);
+
+        private final Set<String> usernames;
+        private final String realmName;
+        private final String realmType;
+        private final Authentication.AuthenticationType authenticationType;
+
+        Group(Set<String> usernames) {
+            this(usernames, null);
+        }
+
+        Group(Set<String> usernames, @Nullable String realmName) {
+            this(usernames, realmName, null, null);
+        }
+
+        Group(Set<String> usernames, @Nullable String realmName, @Nullable String realmType,
+                     @Nullable String authenticationType) {
+            this.usernames = usernames;
+            this.realmName = realmName;
+            this.realmType = realmType == null ? FileRealmSettings.TYPE : realmType;
+            this.authenticationType = authenticationType == null ? Authentication.AuthenticationType.REALM :
+                Authentication.AuthenticationType.valueOf(authenticationType.toUpperCase(Locale.ROOT));
+            validate();
+        }
+
+        private void validate() {
+            final ValidationException validationException = new ValidationException();
+            if (false == FileRealmSettings.TYPE.equals(realmType)) {
+                validationException.addValidationError("[realm_type] only supports [file]");
+            }
+            if (Authentication.AuthenticationType.REALM != authenticationType) {
+                validationException.addValidationError("[auth_type] only supports [realm]");
+            }
+            if (realmName == null) {
+                if (false == SINGLETON_REALM_TYPES.contains(realmType)) {
+                    validationException.addValidationError(
+                        "[realm_name] must be specified for realm types other than [reserved], [file] and [native]");
```

Comment:
This message should use `SINGLETON_REALM_TYPES`

Reply:
Right. Not sure why I used it in the above lines and didn't use it here ..